dlib/matrix/kiss_fft.h

```diff
@@ -0,0 +1,733 @@
+
+        template<typename T>
+        inline kiss_fftndr_state<T>::kiss_fftndr_state(const plan_key& key)
+        {
+            std::vector<int> frontdims = key.dims;
+            frontdims.pop_back();
+            cfg_r  = kiss_fftr_state<T>(plan_key({key.dims.back()}, key.is_inverse));
+            cfg_nd = kiss_fftnd_state<T>(plan_key(frontdims, key.is_inverse));
+        }
+
+        template<typename T>
+        void kiss_fftndr(const kiss_fftndr_state<T>& plan, const T* timedata, std::complex<T>* freqdata)
+        {
+            int dimReal  = plan.cfg_r.substate.nfft*2; //recall the real fft size is half the length of the input
+            int dimOther = plan.cfg_nd.dimprod();
+            int nrbins   = dimReal/2+1;
+
+            std::vector<std::complex<T>> tmp1(std::max<int>(nrbins, dimOther));
+            std::vector<std::complex<T>> tmp2(plan.cfg_nd.dimprod()*dimReal);
+
+            // take a real chunk of data, fft it and place the output at correct intervals
+            for (int k1 = 0; k1 < dimOther; ++k1) 
+            {
+                kiss_fftr(plan.cfg_r, timedata + k1*dimReal , &tmp1[0]); // tmp1 now holds nrbins complex points
+                for (int k2 = 0; k2 < nrbins; ++k2)
+                   tmp2[k2*dimOther+k1] = tmp1[k2];
+            }
+
+            for (int k2 = 0; k2 < nrbins; ++k2) 
+            {
+                kiss_fftnd(plan.cfg_nd, &tmp2[k2*dimOther], &tmp1[0]);  // tmp1 now holds dimOther complex points
+                for (int k1 = 0; k1 < dimOther; ++k1) 
+                    freqdata[ k1*(nrbins) + k2] = tmp1[k1];
+            }
+        }
+
+        template<typename T>
+        void kiss_fftndri(const kiss_fftndr_state<T>& plan, const std::complex<T>* freqdata, T* timedata)
+        {
+            int dimReal  = plan.cfg_r.substate.nfft*2; //recall the real fft size is half the length of the input
+            int dimOther = plan.cfg_nd.dimprod();
+            int nrbins   = dimReal/2+1;
+
+            std::vector<std::complex<T>> tmp1(std::max<int>(nrbins, dimOther));
+            std::vector<std::complex<T>> tmp2(plan.cfg_nd.dimprod()*dimReal);
+
+            for (int k2 = 0; k2 < nrbins; ++k2) 
+            {
+                for (int k1 = 0; k1 < dimOther; ++k1) 
+                    tmp1[k1] = freqdata[ k1*(nrbins) + k2 ];
+                kiss_fftnd(plan.cfg_nd, &tmp1[0], &tmp2[k2*dimOther]);
+            }
+
+            for (int k1 = 0; k1 < dimOther; ++k1) 
+            {
+                for (int k2 = 0; k2 < nrbins; ++k2)
+                    tmp1[k2] = tmp2[ k2*dimOther+k1 ];
+                kiss_fftri(plan.cfg_r, &tmp1[0], timedata + k1*dimReal);
+            }
+        }
+
+        struct hasher
+        {
+            size_t operator()(const plan_key& key) const {return key.hash();}
+        };
+
+        template<typename plan_type>
+        class cache
+        {
+        public:
+            static cache& get()
+            {
+                static cache singleton;
```

Comment:
Might be faster to use `thread_local` instead of `static` and then not use the mutex.

Reply:
i've never used ```thread_local``` before. I would need to read up on it. I could do without the mutex if i used ```thread_local``` ??

Reply:
Ooof. Not sure i like ```thread_local```. @davisking is it ok to leave the mutex? I'm a bit nervous using ```thread_local```. Maybe using that could be another PR which does some performance measurements with and without it. I imagine there will be other PRs relating to FFTs in the near future that add SIMD optimisations for example. So it's not necessary to cram everything into this 1 PR.

Reply:
Also, if you were using ```dlib::parallel_for``` for example, with a fresh ```dlib::thread_pool```, and each thread was doing some FFTs, and you were using ```thread_local```, you would have to recreate the plans every time. I feel like using mutex should be better. But again, i could be proven wrong in a future PR that tests this more thoroughly.

Reply:
Or could use a spinlock instead.

Reply:
A mutex is fine. `thread_local` gives you the potential to reduce contention among many threads though.  But you should measure to see what is best.  In any case, using a mutex here is fine.

This code could also be made a bit easier to follow by using a single templated function that returned a reference to the plan.  Like:
```
template <typename plan_type>
plan_type& get_plan(const plan_key& key)
{
    static std::mutex m;
    static std::unordered_map<plan_key, plan_type, hasher> plans;
    ...
}
```

Reply:
I hadn't thought of doing it that way. Very nice. Makes it a bit more functional.